pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -694,27 +688,34 @@ func (vs *VSphere) CreateVolume(volumeOptions *vclib.VolumeOptions) (volumePath
 			glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
 			return "", err
 		}
-		volumePath = kubeVolsPath + volumeOptions.Name + ".vmdk"
+		volumePath := kubeVolsPath + volumeOptions.Name + ".vmdk"
 		disk := diskmanagers.VirtualDisk{
 			DiskPath:      volumePath,
 			VolumeOptions: volumeOptions,
 			VMOptions:     vmOptions,
 		}
-		err = disk.Create(ctx, ds)
+		volumePath, err = disk.Create(ctx, ds)
 		if err != nil {
 			glog.Errorf("Failed to create a vsphere volume with volumeOptions: %+v on datastore: %s. err: %+v", volumeOptions, datastore, err)
 			return "", err
 		}
+		// Get the canonical path for the volume path.
+		canonicalVolumePath, err = getcanonicalVolumePath(ctx, dc, volumePath)
+		if err != nil {
+			glog.Errorf("Failed to get canonical vsphere volume path for volume: %s with volumeOptions: %+v on datastore: %s. err: %+v", volumePath, volumeOptions, datastore, err)
+			return "", err
+		}
 		if filepath.Base(datastore) != datastore {
 			// If datastore is within cluster, add cluster path to the volumePath
-			volumePath = strings.Replace(volumePath, filepath.Base(datastore), datastore, 1)
+			canonicalVolumePath = strings.Replace(canonicalVolumePath, filepath.Base(datastore), datastore, 1)
 		}
-		return volumePath, nil
+		return canonicalVolumePath, nil
 	}
 	requestTime := time.Now()
-	volumePath, err = createVolumeInternal(volumeOptions)
+	canonicalVolumePath, err = createVolumeInternal(volumeOptions)
 	vclib.RecordCreateVolumeMetric(volumeOptions, requestTime, err)
-	return volumePath, err
+	glog.V(1).Infof("The canonical volume path for the newly created vSphere volume is %q", canonicalVolumePath)
+	return canonicalVolumePath, err
```

Comment:
@BaluDontu 
We are now returning canonicalVolumePath from this function. We have some E2E test cases (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere_volume_placement.go), which is using this function and building PV spec from the returned path. Have you verified that the tests are passing with this change?

Reply:
Yes @divyenpatel . I have verified it and all the tests pass.